@DeepDiver1975 this time it wasn't me:
```

12:18:09 PhantomJS 2.1.1 (Linux 0.0.0): Executed 835 of 838 (skipped 3) ERROR (21.735 secs / 20.435 secs)
12:18:11 [33m06 03 2017 12:18:11.133:WARN [launcher]: [39mPhantomJS was not killed in 2000 ms, sending SIGKILL.
12:18:11 Makefile:185: recipe for target 'test-js' failed
12:18:11 make: *** [test-js] Error 1
```

--- a/lib/private/Files/Cache/Storage.php
+++ b/lib/private/Files/Cache/Storage.php
@@ -85,14 +85,14 @@ class Storage {
 	/**
 	 * query the local cache, a distributed cache and the db for a storageid
 	 * @param string $storageId
-	 * @return array|null
+	 * @return array|false
 	 */
 	public static function getStorageById($storageId) {
 		if (self::$localCache === null) {
 			self::$localCache = new CappedMemoryCache();
 		}
 		$result = self::$localCache->get($storageId);
-		if (empty($result)) {
+		if ($result === null) {
 			$result = self::getStorageByIdFromCache($storageId);
 			self::$localCache->set($storageId, $result);
 		}
@@ -102,7 +102,7 @@ class Storage {
 	/**
 	 * query the distributed cache for a storageid
 	 * @param string $storageId
-	 * @return array|null
+	 * @return array|false
 	 */
 	private static function getStorageByIdFromCache($storageId) {
 		if (self::$distributedCache === null) {
@@ -110,7 +110,7 @@ class Storage {
 				\OC::$server->getMemCacheFactory()->create('getStorageById');
 		}
 		$result = self::$distributedCache->get($storageId);
-		if (empty($result)) {
+		if ($result === null) {
 			$result = self::getStorageByIdFromDb($storageId);
 			self::$distributedCache->set(
 				$storageId,
@@ -124,7 +124,7 @@ class Storage {
 	/**
 	 * query the db for a storageid
 	 * @param string $storageId
-	 * @return array|null
+	 * @return array|false
 	 */
 	private static function getStorageByIdFromDb($storageId) {
 		$sql = 'SELECT * FROM `*PREFIX*storages` WHERE `id` = ?';
@@ -232,7 +232,7 @@ class Storage {
 		\OC_DB::executeAudited($sql, [$storageId]);
 
 		// delete from local cache
-		if(self::$localCache) {
+		if(self::$localCache === null) {
 			self::$localCache->remove($storageId);
 		}
 		// delete from distributed cache

--- a/lib/private/Files/Cache/Storage.php
+++ b/lib/private/Files/Cache/Storage.php
@@ -26,6 +26,8 @@
  */
 
 namespace OC\Files\Cache;
+use OC\Cache\CappedMemoryCache;
+use OCP\ICache;
 
 /**
  * Handle the mapping between the string and numeric storage ids
@@ -41,6 +43,14 @@ namespace OC\Files\Cache;
 class Storage {
 	private $storageId;
 	private $numericId;
+
+	/** @var CappedMemoryCache */
+	protected static $localCache = null;
+
+	/** @var ICache  */
+	protected static $distributedCache = null;
+
+	protected static $distributedCacheTTL = 300; // 5 Min
 
 	/**
 	 * @param \OC\Files\Storage\Storage|string $storage
@@ -73,13 +83,53 @@ class Storage {
 	}
 
 	/**
+	 * query the local cache, a distributed cache and the db for a storageid
 	 * @param string $storageId
 	 * @return array|null
 	 */
 	public static function getStorageById($storageId) {
+		if (self::$localCache === null) {
+			self::$localCache = new CappedMemoryCache();
+		}
+		$result = self::$localCache->get($storageId);
+		if (empty($result)) {
+			$result = self::getStorageByIdFromCache($storageId);
+			self::$localCache->set($storageId, $result);
+		}
+		return $result;
+	}
+
+	/**
+	 * query the distributed cache for a storageid
+	 * @param string $storageId
+	 * @return array|null
+	 */
+	private static function getStorageByIdFromCache($storageId) {
+		if (self::$distributedCache === null) {
+			self::$distributedCache =
+				\OC::$server->getMemCacheFactory()->create('getStorageById');
+		}
+		$result = self::$distributedCache->get($storageId);
+		if (empty($result)) {
+			$result = self::getStorageByIdFromDb($storageId);
+			self::$distributedCache->set(
+				$storageId,
+				$result,
+				self::$distributedCacheTTL
+			);
+		}
+		return $result;
+	}
+
+	/**
+	 * query the db for a storageid
+	 * @param string $storageId
+	 * @return array|null
+	 */
+	private static function getStorageByIdFromDb($storageId) {
 		$sql = 'SELECT * FROM `*PREFIX*storages` WHERE `id` = ?';
-		$result = \OC_DB::executeAudited($sql, [$storageId]);
-		return $result->fetchRow();
+		$resultSet = \OC_DB::executeAudited($sql, [$storageId]);
+		return $resultSet->fetchRow();
 	}
 
 	/**
@@ -181,6 +231,17 @@ class Storage {
 		$sql = 'DELETE FROM `*PREFIX*storages` WHERE `id` = ?';
 		\OC_DB::executeAudited($sql, [$storageId]);
 
+		// delete from local cache
+		if(self::$localCache) {
+			self::$localCache->remove($storageId);
+		}
+		// delete from distributed cache
+		if (self::$distributedCache === null) {
+			self::$distributedCache =
+				\OC::$server->getMemCacheFactory()->create('getStorageById');
+		}
+		self::$distributedCache->remove($storageId);
+		// delete from db
 		if (!is_null($numericId)) {
 			$sql = 'DELETE FROM `*PREFIX*filecache` WHERE `storage` = ?';
 			\OC_DB::executeAudited($sql, [$numericId]);

--- a/lib/private/Files/Cache/Storage.php
+++ b/lib/private/Files/Cache/Storage.php
@@ -135,7 +135,7 @@ class Storage {
 
 	private static function unsetCache($storageId) {
 		// delete from local cache
-		if(self::$localCache === null) {
+		if(self::$localCache !== null) {
 			self::$localCache->remove($storageId);
 		}
 		// delete from distributed cache
@@ -219,6 +219,8 @@ class Storage {
 	 * @param bool $isAvailable
 	 */
 	public function setAvailability($isAvailable) {
+		// delete from local cache
+		self::unsetCache($this->storageId);
 		$sql = 'UPDATE `*PREFIX*storages` SET `available` = ?, `last_checked` = ? WHERE `id` = ?';
 		$available = $isAvailable ? 1 : 0;
 		\OC_DB::executeAudited($sql, [$available, time(), $this->storageId]);

